src/config/compiler.rs

```diff
@@ -0,0 +1,80 @@
+use super::{builder::ConfigBuilder, validation, Config, TransformOuter};
+use indexmap::IndexMap;
+
+pub fn compile(raw: ConfigBuilder) -> Result<Config, Vec<String>> {
+    let mut config = Config {
+        global: raw.global,
+        sources: raw.sources,
+        sinks: raw.sinks,
+        transforms: raw.transforms,
+        tests: raw.tests,
+        expansions: Default::default(),
+    };
+
+    let mut errors = Vec::new();
+
+    if let Some(warnings) = validation::warnings(&config) {
+        for warning in warnings {
+            warn!(message = %warning)
+        }
+    }
+
+    if let Err(type_errors) = validation::check_shape(&config) {
+        errors.extend(type_errors);
+    }
+
+    if let Err(type_errors) = validation::typecheck(&config) {
+        errors.extend(type_errors);
+    }
+
+    expand_macros(&mut config)?;
+
+    if errors.is_empty() {
+        Ok(config)
+    } else {
+        Err(errors)
+    }
+}
+
+/// Some component configs can act like macros and expand themselves into
+/// multiple replacement configs. Returns a map of components to their
```

Comment:
This suggests it returns the map, but it sets it in `config.expansions` instead.